dv/verilator/memutil/cpp/verilator_memutil.cc

```diff
@@ -0,0 +1,432 @@
+// Copyright lowRISC contributors.
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+
+#include "verilator_memutil.h"
+
+#include <fcntl.h>
+#include <gelf.h>
+#include <getopt.h>
+#include <libelf.h>
+#include <sys/stat.h>
+#include <unistd.h>
+
+#include <cstring>
+#include <iostream>
+#include <list>
+
+// DPI Exports
+extern "C" {
+
+/**
+ * Write |file| to a memory
+ *
+ * @param file path to a SystemVerilog $readmemh()-compatible file (VMEM file)
+ */
+extern void simutil_verilator_memload(const char *file);
+
+/**
+ * Write a 32 bit word |val| to memory at index |index|
+ *
+ * @return 1 if successful, 0 otherwise
+ */
+extern int simutil_verilator_set_mem(int index, const svLogicVecVal *val);
+}
+
+struct BufferDesc {
+  uint8_t *data;
+  size_t length;
+};
+
+bool VerilatorMemUtil::RegisterMemoryArea(const std::string name,
+                                          const std::string location) {
+  MemArea mem = {.name = name, .location = location};
+
+  auto ret = mem_register_.emplace(name, mem);
+  if (ret.second == false) {
+    std::cerr << "ERROR: Can not register \"" << name << "\" at: \"" << location
+              << "\" (Previously defined at: \"" << ret.first->second.location
+              << "\")" << std::endl;
+    return false;
+  }
+  return true;
+}
+
+bool VerilatorMemUtil::ParseCommandArgs(int argc, char **argv, bool &exit_app) {
+  const struct option long_options[] = {
+      {"rominit", required_argument, nullptr, 'r'},
+      {"raminit", required_argument, nullptr, 'm'},
+      {"flashinit", required_argument, nullptr, 'f'},
+      {"meminit", required_argument, nullptr, 'l'},
+      {"help", no_argument, nullptr, 'h'},
+      {nullptr, no_argument, nullptr, 0}};
+
+  // Reset the command parsing index in-case other utils have already parsed
+  // some arguments
+  optind = 1;
+  while (1) {
+    int c = getopt_long(argc, argv, ":r:m:f:l:h", long_options, nullptr);
```

Comment:
Probably not something that has to be done for this PR but it'd be nice to purge the C-style argument list handling